### ISSUE TITLE
Performance problems in repo with lots of submodules

### ISSUE THREAD

**Issue by danielmenzel** [gitextensions/gitextensions#5733]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

When opening a repository with lots of submodules (65 in our case) GitExtensions hangs for about a minute with no UI updates. This then also happens repeatedly when working inside the opened repository. I captured a Git Command Log while opening the repository:

![gitext_log_slow_submodules](https://user-images.githubusercontent.com/8415713/48249279-d504f080-e3fa-11e8-8ef9-f378e2e35371.PNG)

During the hang nothing happens in the main window or the log window (the mouse cursor changes to the "waiting circle"), then all submodule commands show up at the same time in the log window.

I also profiled a GitExtensions release build  #(89897d085a60823f72ba5045fad6d84dd02bc58b) in VisualStudio which seems to indicate that waiting for all submodule status update tasks in [GetRepositorySubmodulesStatusAsync](https://github.com/gitextensions/gitextensions/blob/c2eb7fee8433d0ead1ca3eb5d0c0df87211f53b5/GitCommands/Submodules/SubmoduleStatusProvider.cs#L99) takes a lot of time:

![gitext_perf](https://user-images.githubusercontent.com/8415713/48249836-8bb5a080-e3fc-11e8-9ff8-87ac513b94c0.PNG)

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

- Open a repository with lots of submodules

**What is the expected behavior?**

UI should not hang when opening a repository

**Environment you encounter the issue:**

- Git Extensions 3.00.00.02-rc1
- 7990325547dd27e6a3fc1eaccee2607bbce66222  (Dirty)
- Git 2.19.1.windows.1
- Microsoft Windows NT 10.0.16299.0
- .NET Framework 4.7.3190.0

**Did this work in previous version of GitExtensions (which)?**

2.51.05 is also not ideal in this case (some shorter hangs with lots of submodule related activity in the log window) but does not hang for such a long period of time with no UI updates.

**Comment by RussKie** [comment]
@gerhardol @sharwell It would be great if you be able to look into this

**Comment by gerhardol** [comment]
Not this problem, but an explanation of a change in 3.0:
In 2.x, the submodule status (and stash count) was updated every time the Browse window was activated. In 3.0 the update is done at changes of submodules (using the commit count), as well throttled to once every 15th sec.
I believe GetRepositorySubmodulesStatusAsync is for the update of status in the menu, probably not related to the slowdown seen.

I can see the spinning occasionally longer in 3.0, but it does not seem to be related to number of submodules but more to repo size.

**Comment by spdr870** [comment]
@gerhardol I checked, and in some repo's with submodules I noticed the slower loading. I have the status in the menu disabled, because this is too slow for me. Still GetRepositorySubmodulesStatusAsync is getting executed. It should be async, but its seems to be blocking something. The larger the submodules, the longer the slowdown.

Normally the first commits are shown really quick. It should be faster than in 2.x. I did notice this:
- All refs are loaded before starting to read the log. When you have a lot of refs (I sometimes have 12k) this causes a slowdown.
- Submodules are havy on IO, since it checks for changes in each submodule. Even if your not interested in them.
- When the tree is enabled, this switches to the UI tread a lot. Usually not that slow, except when you have a lot of local branches.

**Comment by gerhardol** [comment]
> Still GetRepositorySubmodulesStatusAsync is getting executed. 

Yes, only once though.
Should be tested by disabling the submodule status. This started before the GetRepositorySubmodulesStatusAsync change in August, maybe at the threading changes in March.

**Comment by RussKie** [comment]
I'm not sure if it is the same issue, but feels related

I don't have many submodules, but I see a significant slow down if I have the left panel open.
Every time the working dir state changes, the app pauses...

That's what the log shows:
![image](https://user-images.githubusercontent.com/4403806/48288538-8faaf680-e4c0-11e8-9875-43721b4f4856.png)

with the left panel open:
![left-panel-slow1](https://user-images.githubusercontent.com/4403806/48288564-b0734c00-e4c0-11e8-939d-c47b9ebe4546.gif)

with the left panel closed:
![left-panel-slow2](https://user-images.githubusercontent.com/4403806/48288762-4a3af900-e4c1-11e8-9176-8d1dcd52ac10.gif)

**Comment by gerhardol** [comment]
I have tried adding GE as a submodule to itself to get over 30 submodules to reproduce. Sometimes it is OK, sometimes it is slower.
One reason though is that the cache for git common dir is not working as intended, it has no lock. For me, the first "git rev-parse --git-common-dir" command takes about 30 ms, as they are queued, the last can take over 350 ms.
As many GitModules are created about the same time, the rev-parse command can still require >200ms, but there are less calls. Maybe a common cache is possible. The PR can definitely be improved.

**Comment by gerhardol** [comment]
Please try RC2: https://github.com/gitextensions/gitextensions/releases/tag/v3.00.00-rc2
I can notice an improvement, not all though.

**Comment by danielmenzel** [comment]
I installed RC2 but it is still the same behavior (even with the left sidebar disabled): about a minute hang with total UI freeze upon opening the repository an then repeated 1 minute freezes while trying to work inside the repo. For example selecting a different commit sometimes triggers the hang or even opening the file menu.

When I pause program execution with the debugger during the hang this is the callstack of the Main Thread:

```
 	[External Code]	
>	GitCommands.dll!GitCommands.ExecutableExtensions.GetOutput(GitCommands.IExecutable executable, GitCommands.ArgumentString arguments, byte[] input, System.Text.Encoding outputEncoding, GitCommands.CommandCache cache, bool stripAnsiEscapeCodes) Line 46	C#
 	GitCommands.dll!GitCommands.GitModule.GetDiffFiles(string firstRevision, string secondRevision, string parentToSecond, bool noCache) Line 2470	C#
 	GitCommands.dll!GitCommands.GitModule.GetDiffFilesWithSubmodulesStatus(string firstRevision, string secondRevision, string parentToSecond) Line 2461	C#
 	GitUI.dll!GitUI.FileStatusList.SetDiffs(System.Collections.Generic.IReadOnlyList<GitCommands.GitRevision> revisions) Line 608	C#
 	GitUI.dll!GitUI.CommandsDialogs.RevisionDiffControl.DisplayDiffTab() Line 145	C#
 	GitUI.dll!GitUI.CommandsDialogs.FormBrowse.FillDiff() Line 1184	C#
 	GitUI.dll!GitUI.CommandsDialogs.FormBrowse..ctor.AnonymousMethod__55_1(object sender, System.EventArgs e) Line 281	C#
 	GitUI.dll!GitUI.RevisionGridControl..ctor.AnonymousMethod__109_0(object _, System.EventArgs e) Line 161	C#
 	[External Code]	
 	GitUI.dll!GitUI.GitUICommands.StartBrowseDialog(System.Windows.Forms.IWin32Window owner, string filter, GitUIPluginInterfaces.ObjectId selectedCommit) Line 1076	C#
 	GitUI.dll!GitUI.GitUICommands.RunBrowseCommand(System.Collections.Generic.IReadOnlyList<string> args) Line 1549	C#
 	GitUI.dll!GitUI.GitUICommands.RunCommandBasedOnArgument(System.Collections.Generic.IReadOnlyList<string> args, System.Collections.Generic.IReadOnlyDictionary<string, string> arguments) Line 1381	C#
 	GitUI.dll!GitUI.GitUICommands.RunCommand(System.Collections.Generic.IReadOnlyList<string> args) Line 1350	C#
 	GitExtensions.exe!GitExtensions.Program.RunApplication() Line 140	C#
 	GitExtensions.exe!GitExtensions.Program.Main() Line 64	C#
```

**Comment by vbjay** [comment]
Try running git gc --aggressive

**Comment by vbjay** [comment]
Also turn off showing super project branches in the grid. See if that does anything.

**Comment by gerhardol** [comment]
@danielmenzel Can you please test the portable build for #5747 (or build it yourself) to see if this solves your problem?

If that works, the solution for 3.0 should be to turn off submodule status by default and make it configurable.

I have tried to limit the commands:
 * Creation of GitModule can be significantly reduced in GetSubmoduleStatusAsync() (very hacky solutions so far). For instance when making copies, all existing structures can be copied from the existing module. However, you still have a slowdown of the updates, so not solving the core problem, just improving.
 * Delaying creating the threads will help, too, but this will take a very long time with many submodules to be safe if the time is fixed. One solution is to await the commit-count update and only update the submodules with changes. May still be slow.

Other notes: 
* Some RevisionDiff commands seem to run several times
* Will Thread priority help here?
* Status is only seem to be calculated one level down, should be recursive

**Comment by gerhardol** [comment]
Thanks for testing
Then I propose that we add a setting for Submodule status for 3.0. This will require translation updates too.
Optimize the submodule status is then not a priority for 3.0.

**Comment by RussKie** [comment]
Thank you Gerhard.

> This will require translation updates too.

That's ok, don't stress about it.

**Comment by amaiorano** [comment]
@gerhardol Can we try to seeif @danielmenzel's slow down would disappear by _only_ disabling GetSubmoduleStatusAsync? In other words, without all the delays.

**Comment by amaiorano** [comment]
If I understand correctly, the problem of the UI not updating for so long on startup/repo switch/refresh has to do with the fact that FormBrowse is waiting on certain git operations to complete before it can render its graph (and ROT as well), and during this time, there's a whole lot of submodule updates (e.g. git -c diff.submodule=short -c diff.noprefix=false diff --no-color -- <submodule name>) that have been started. So is the problem that git commands like these can only really run one at a time? That's the choke point? We've got this long queue of git commands, and the less crucial submodule ones are running before the more crucial current repo ones?

**Comment by gerhardol** [comment]
@amaiorano 
> by only disabling GetSubmoduleStatusAsync? In other words, without all the delays.

Of course, if @danielmenzel agrees. I see major slowdowns occasionally when testing with 30+ repos on a GE branch too, so it is not strictly required, I assume.

However, I doubt the "minute hanging" is seen without the delays. There could however be "seconds hanging" that are harder to investigate. As the delayed info is not needed immediately the delay should not cause noticeable delays. 

Background commands should use "--no-optional-locks" which may help for instance.
Really hope that a GitCommandRunner service will not be needed.

I can submit two PRs so the delay or submodule-status-configuration can be voted down or if someone find a better solution. Current prio is to release 3.0 safely so we can release the "lockdown" of features in master, for instance for #5669

**Comment by gerhardol** [comment]
> Can we try putting just one delay in FormBrowse.InitiateSubmodulesUpdate just before we invoke _submoduleStatusProvider.UpdateSubmodulesStatus?

Then Browse will have to create the thread first.

> What's stopping us from using "--no-optional-locks"?

#5067 implemented this for fit-status for commit count.
I did not expect (and still do not understand) why git-diff would use a lock though.

I will be happy if you are looking into if this or something else improves the performance. I will likely be home late tonight anyway and should prepare a "safe" PR anyway.

New devs are welcome to apply.

**Comment by amaiorano** [comment]
> Then Browse will have to create the thread first.

Right, something along the lines of:

```csharp
        private void InitiateSubmodulesUpdate()
        {
            if (!_submoduleStatusUpdateNeeded)
            {
                return;
            }

            toolStripButtonLevelUp.ToolTipText = "";
            _submoduleStatusUpdateNeeded = false;

            // HACK: Delay submodule status update so that it doesn't content with module status update
            // used to refresh FormBrowse, etc.
            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
            {
                await Task.Delay(100);
                _submoduleStatusProvider.UpdateSubmodulesStatus(Module.WorkingDir, _noBranchTitle.Text);
            }).FileAndForget();
        }
```

> > What's stopping us from using "--no-optional-locks"?
> 
> #5067 implemented this for fit-status for commit count.
> I did not expect (and still do not understand) why git-diff would use a lock though.

Right, I guess we can look at git source. But isn't this something easy enough to try? Just see if it makes a difference?
 
> I will be happy if you are looking into if this or something else improves the performance. I will likely be home late tonight anyway and should prepare a "safe" PR anyway.
> 
> New devs are welcome to apply.

I would love to, but can't at the moment. Hopefully some time will free up soon enough. Also, keeping submodule status updates for is near and dear to my heart as it's something I use and want in the Submodules ROT change I made :)

**Comment by gerhardol** [comment]
I would not call the delay a hack. It is a good approach to start with what is important.
Threads have priority too.

> Right, I guess we can look at git source. But isn't this something easy enough to try? Just see if it makes a difference?

Of course, just test. It is a lot more likely that it works OK than it finds the cure for cancer.

**Comment by RussKie** [comment]
> What's stopping us from using "--no-optional-locks"? Just read about it, and it sounds great. I see that #5066, which was exactly about, was closed, but I don't know why.

You fixed in #5067 😎 

> I would prefer this myself, but I'm not a core dev.

We're all volunteers here, so everyone has the same right to express their opinions. The core team then bears responsibility for what we choose to implement.
Be careful what you wish for, you wish may be granted 😎

**Comment by amaiorano** [comment]
> > What's stopping us from using "--no-optional-locks"? Just read about it, and it sounds great. I see that #5066, which was exactly about, was closed, but I don't know why.
> 
> You fixed in #5067 😎

You mean @gerhardol fixed it, right?

> > I would prefer this myself, but I'm not a core dev.
> 
> We're all volunteers here, so everyone has the same right to express their opinions. The core team then bears responsibility for what we choose to implement.

Your comment implies there is a core team, and that you're not part of it? I thought you, @gerhardol , @spdr870, and others I've seen opening and merging PRs made up the core team...

> Be careful what you wish for, you wish may be granted 😎

I'm definiteley having fun working on GE with y'all so far :)

**Comment by amaiorano** [comment]
> I would not call the delay a hack. It is a good approach to start with what is important.
> Threads have priority too.

I don't mean to belabor the point, but I do think it's a hack to use delays to manage thread contention because it's non-deterministic. This is the reason I'd rather see no delays anywhere in the submodule status provider code, and only in FormBrowse when we launch the thread, because it localizes the hack to one place, and it's the place where the other threads are started, so it's at least a little easier to reason about.

**Comment by danielmenzel** [comment]
> Of course, if @danielmenzel agrees.

I don't mind testing other solutions.

**Comment by gerhardol** [comment]
> This is the reason I'd rather see no delays anywhere in the submodule status provider code, and only in FormBrowse when we launch the thread

That part is a hack, I made a minimal PR so the delay is in the provider right now

> You fixed in #5067

That was only for git-status for commit-count, may be needed on more commands

> No problem, I can also test other solutions.

#5756 has the configuration for submodule status
#5747 is only for the delay

**Comment by danielmenzel** [comment]
#5756 works as expected: `submodule status` disabled -> no hang; enabled -> hang

#5747 (build 4172) hangs

**Comment by gerhardol** [comment]
I opened a new issue with suggested way forward with the thoughts I have.
#5747 (as well as #5742) do not make a major difference, but I still believe they are improvements.
